src/Uno.UWP/UI/Composition/Uno/CompositorThread.android.cs

```diff
@@ -0,0 +1,186 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.Foundation;
+using Android.App;
+using Android.Graphics;
+using Android.OS;
+using Android.Views;
+using Uno.Extensions;
+using Uno.Logging;
+using Exception = System.Exception;
+using Point = Windows.Foundation.Point;
+using Rect = Windows.Foundation.Rect;
+using ThreadPriority = System.Threading.ThreadPriority;
+
+namespace Uno.UI.Composition
+{
+	internal interface ICompositionRoot
+	{
+		Window Window { get; }
+
+		View Content { get; }
+	}
+
+	internal class CompositorThread : Java.Lang.Object, ISurfaceHolderCallback2, Choreographer.IFrameCallback
+	{
+		[ThreadStatic] // Static to the related UI Thread !!!
+		private static CompositorThread? _current;
+
+		public static void Start(ICompositionRoot activity)
+			=> _current ??= new CompositorThread(activity);
+
+		//private RenderNode _animationsNode;
+		//private List<(RenderNode node, double x, double y)> _animatedNodes = new List<(RenderNode node, double x, double y)>();
+
+		private readonly HardwareRenderer _hwRender = new HardwareRenderer();
+		private readonly RenderNode _visualTree = new RenderNode("visual-tree");
+		private readonly ManualResetEventSlim _frameRendered = new ManualResetEventSlim(false);
+		private readonly ICompositionRoot _activity;
+
+		private Thread? _thread;
+
+		public CompositorThread(ICompositionRoot activity)
+		{
+			_activity = activity;
+
+			_hwRender.SetContentRoot(_visualTree);
+			_activity.Window!.TakeSurface(this);
+		}
+		
+		void ISurfaceHolderCallback.SurfaceCreated(ISurfaceHolder? holder)
+		{
+			_hwRender.SetSurface(holder!.Surface);
+
+			Start();
+		}
+
+		void ISurfaceHolderCallback.SurfaceChanged(ISurfaceHolder holder, Format format, int width, int height)
+		{
+			_visualTree.SetPosition(new Rect(new Point(), new Size(width, height)));
+		}
+
+		void ISurfaceHolderCallback.SurfaceDestroyed(ISurfaceHolder? holder)
+		{
+			Stop();
+
+			_hwRender?.Stop();
+			_hwRender?.Destroy(); // Note: This only release the surface, we can restore it by setting a new surface
+		}
+
+		void ISurfaceHolderCallback2.SurfaceRedrawNeeded(ISurfaceHolder? holder)
+		{
+			_frameRendered.Reset();
+			_frameRendered.Wait();
+		}
+
+		void ISurfaceHolderCallback2.SurfaceRedrawNeededAsync(ISurfaceHolder holder, Java.Lang.IRunnable drawingFinished)
+		{
+			_frameRendered.Reset();
+
+			Task.Run(() => // TODO: Should we invoke drawingFinished on calling / UI thread?
+			{
+				try
+				{
+					_frameRendered.Wait();
+					drawingFinished.Run();
+				}
+				catch (Exception e)
+				{
+					this.Log().Error("Async redraw failed.", e);
+				}
+			});
+		}
+
+		private void Start()
+		{
+			if (_thread is {})
+			{
+				this.Log().Warn("Try to start an already running compositor thread.");
+				return;
+			}
+
+			_thread = new Thread(() =>
+			{
+				_hwRender.Start();
+
+				Looper.Prepare(); // Required to get a Choreographer.Instance
+				Choreographer.Instance!.PostFrameCallback(this);
```

Comment:
Is it really always not null?

Reply:
Yes ... as long as we have previously called the `Looper.Prepare()`.